support/cas-server-support-x509-core/src/main/java/org/apereo/cas/adaptors/x509/authentication/principal/X509UPNExtractor.java

```diff
@@ -0,0 +1,124 @@
+package org.apereo.cas.adaptors.x509.authentication.principal;
+
+import org.apereo.cas.util.function.FunctionUtils;
+
+import com.google.common.base.Predicates;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+
+import org.bouncycastle.asn1.ASN1InputStream;
+import org.bouncycastle.asn1.ASN1ObjectIdentifier;
+import org.bouncycastle.asn1.ASN1OctetString;
+import org.bouncycastle.asn1.ASN1Sequence;
+import org.bouncycastle.asn1.ASN1String;
+import org.bouncycastle.asn1.ASN1TaggedObject;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.security.cert.CertificateParsingException;
+import java.security.cert.X509Certificate;
+import java.util.List;
+
+/**
+ * Credential to principal resolver that extracts Subject Alternative Name UPN extension
+ * from the provided certificate if available as a resolved principal id.
+ *
+ * @author Dmitriy Kopylenko
+ * @since 4.1.0
+ */
+@Slf4j
+public class X509UPNExtractor {
```

Comment:
This is a utility class, and should be marked with @UtilityClass; the name probably should end with `Utils` until perhaps we find a better home for it :)

Reply:
I left static on the class variables even though UtilityClass doesn't require it because checkstyle was complaining that name of what it saw as non-static variable was all caps.